pkg/app/context.go

```diff
@@ -1059,19 +1059,57 @@ func (ctx *RequestContext) Body() ([]byte, error) {
 type ClientIP func(ctx *RequestContext) string
 
 var defaultClientIP = func(ctx *RequestContext) string {
-	RemoteIPHeaders := []string{"X-Real-IP", "X-Forwarded-For"}
-	for _, headerName := range RemoteIPHeaders {
-		ip := ctx.Request.Header.Get(headerName)
-		if ip != "" {
-			return ip
+	RemoteIPHeaders := []string{"X-Forwarded-For", "X-Real-IP"}
+	TrustedProxies := []string{"0.0.0.0"}
+
+	remoteIP, _, err := net.SplitHostPort(strings.TrimSpace(ctx.RemoteAddr().String()))
+	if err != nil {
+		return ""
+	}
+	trusted := isTrustedProxy(TrustedProxies, remoteIP)
+
+	if trusted {
+		for _, headerName := range RemoteIPHeaders {
+			ip, valid := validateHeader(TrustedProxies, ctx.Request.Header.Get(headerName))
+			if valid {
+				return ip
+			}
 		}
 	}
 
-	if ip, _, err := net.SplitHostPort(strings.TrimSpace(ctx.RemoteAddr().String())); err == nil {
-		return ip
+	return remoteIP
+}
+
+// isTrustedProxy will check whether the IP address is included in the trusted list according to TrustedProxies
+func isTrustedProxy(trustedProxies []string, remoteIP string) bool {
+	for _, proxy := range trustedProxies {
+		if proxy == remoteIP {
+			return true
+		}
 	}
+	return false
+}
 
-	return ""
+// validateHeader will parse X-Forwarded-For and X-Real-IP header and return the Initial client IP address or an untrusted IP address
+func validateHeader(trustedProxies []string, header string) (clientIP string, valid bool) {
+	if header == "" {
+		return "", false
+	}
+	items := strings.Split(header, ",")
+	for i := len(items) - 1; i >= 0; i-- {
+		ipStr := strings.TrimSpace(items[i])
+		ip := net.ParseIP(ipStr)
+		if ip == nil {
```

Comment:
seems not work properly and still vulnerable
if hacker add header like this "curl -H 'X-Forwarded-For: x.x.x.x,'" ClientIP() always return ctx.RemoteAddr() since ip == nil in validateHeader()

Reply:
i got it, thx

Reply:
If the intranet server is trusted server and hacker add header like this "curl -H 'X-Forwarded-For: x.x.x.x'" (a fake ip), `ClientIP()` always return the fake ip. this is not a problem that this pr can solve. If the intranet server is untrusted server, `ClientIP()` will return the server's ip correctly